### ISSUE TITLE
[SUGGESTION] BsonDiscriminator and requesting instances of defined type

### ISSUE THREAD

**Issue by frogcrush** [litedb-org/LiteDB#1714]
**Is your feature request related to a problem? Please describe.**
Currently I'm trying to develop an application that uses abstract classes in its POCOs. For example, you would have an abstract class Animal, and the classes Dog and Cat would inherit from Animal. 

Currently, I cannot find a way to do something like follows:

```
db.GetCollection<Animal>()
```

Allows storing items. However,

```
db.GetCollection<Dog>()
```
Does not work, as it returns no items. 

Therefore, I used a named collection:

```
db.GetCollection<Animal>("animals");
// and
db.GetCollection<Dog>("animals");
```
both work correctly until you insert a Cat into the collection. Then it will throw an exception.

**Describe the solution you'd like**
- Add a way to map classes based on a discriminator (like MongoDB's ``BsonDiscriminator``) 
- Add a way to fetch objects of a certain type from a collection (``GetCollection<Animal>().OfType<Dog>()`` for example)

**Describe alternatives you've considered**
Manually writing BSON documents for everything, which defeats the purpose of using POCOs in the first place

A related StackOverflow answer for MongoDB is here: https://stackoverflow.com/questions/10405887/inheritance-in-mongodb-how-to-request-instances-of-defined-type

Entity Framework Core also has a similar feature; https://docs.microsoft.com/en-us/ef/core/modeling/inheritance

**Comment by lbnascimento** [comment]
@frogcrush You should have no issues creating a collection of `Animal`s and inserting instances of `Cat` and `Dog` in it, LiteDB will deserialize the objects to their correct types. There is no specific functionality for filtering the objects by type, but you can easily do it with Linq after retrieval.

**Comment by frogcrush** [comment]
Would filtering objects by type with Linq not have a performance penalty though? 

If so, would there be less penalty using an SQL statement to limit based on the _type column that is generated?

**Comment by lbnascimento** [comment]
@frogcrush Yes, it should be faster to filter over the `_type` field, however this field is not always present. It is only present when the type of a object being inserted is different from the type argument for the collection (which will always be the case in your example, given that `Animal` is an abstract class). And this consistency is only kept if the collection is always opened with a type argument – if insertions are ever made to the collection with untyped collections or with SQL commands, there's no way to guarantee that all the documents in the collection are "correctly typed".

**Comment by JKamsker** [comment]
Thanks for the suggestion! It's now part of #2831, which collects the related feature requests in one place so they can be discussed and prioritised together. Closing this as a duplicate — please follow #2831 and add your use case there.
